core/queueref_test.go

```diff
@@ -56,3 +61,129 @@ func TestQueueRef(t *testing.T) {
+func TestRemoveMultipleExpiredBeforeNonExpire(t *testing.T) {
+	qr := core.NewQueueRef()
+	val := [3]int{10, 20, 30}
+	core.Put("key1", core.NewObj(val[0], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key1")
+	core.Put("key2", core.NewObj(val[1], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key2")
+	core.Put("key3", core.NewObj(val[2], -1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key3")
+	time.Sleep(2 * time.Millisecond)
+	fmt.Printf("Queue size : %d\n", qr.Length())
+	qe, err := qr.Remove()
+	assert.Check(t, err == nil || qe.Obj.Value == val[2], fmt.Sprintf("test for removing mulitple expired key before non-expired failed , Expected : %d, Got %d\n", val[2], qe.Obj.Value))
+}
+
+// Test for removing from queue with all expired keys
+func TestRemoveAllExpired(t *testing.T) {
+	qr := core.NewQueueRef()
+	val := [3]int{10, 20, 30}
+	core.Put("key1", core.NewObj(val[0], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key1")
+	core.Put("key2", core.NewObj(val[1], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key2")
+	core.Put("key3", core.NewObj(val[2], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key3")
+	time.Sleep(2 * time.Millisecond)
+	// remove from empty queue
+	_, err := qr.Remove()
+	assert.Equal(t, err, core.ErrQueueEmpty, fmt.Sprintf("test for removing from empty queue failed Expected : %s, Got : %s\n", core.ErrQueueEmpty, err))
+}
+
+// Benchmark queueref by inserting expired, non-expired and expired keys in order and removing them
+func BenchmarkQueueRef(b *testing.B) {
+	// Setup code...
+	qr := core.NewQueueRef()
+	config.KeysLimit = 100000000 // override keys limit high for benchmarking
+
+	// need to init again for each round with the overriden buffer size
+	// otherwise the watchchannel buffer size will stay as it is with the global keylimits size
+	core.WatchChannel = make(chan core.WatchEvent, config.KeysLimit)
+	b.ResetTimer()
+
+	expiredCount := 0
+	nonExpiredCount := 0
+	b.Run("Insertion of expired and non-expired keys", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			key := fmt.Sprintf("k%d", i)
+			value := fmt.Sprintf("v%d", i)
+			random := rand.Intn(2)
+			// Insert keys without expiration
+			expiration := -1
+			if random == 1 {
+				// Insert keys with expiration of 1ms
+				expiration = 1
+				expiredCount++
+			} else {
+				nonExpiredCount++
+			}
+			core.Put(key, core.NewObj(value, int64(expiration), core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+			qr.Insert(key)
+		}
+	})
+	len := qr.Length()
+	fmt.Printf("Queue length before removal : %d\n", len)
+	fmt.Printf("Inserted keys count, expired : %d , non-expired : %d\n", expiredCount, nonExpiredCount)
+	// Allow keys to expire
+	time.Sleep(2 * time.Millisecond)
+
+	b.Run("Removal of keys", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			// Removal benchmark...
+			_, err := qr.Remove()
+			if err == core.ErrQueueEmpty {
+				break
```

Comment:
B.Fatal here

Reply:
This is a valid case and we are exiting the removal once queue is emptied.

Reply:
Using conditional statements for benchmark control flow is not a good idea.
Benchmarks should display known behaviour. You need to know exactly how many removals you need to do (instead of checking for a breaking condition). Maybe you can restructure the benchmark.
We can probably do that separately.